[WIP] Added three example project templates

mfojtik opened the pull request:
@smarterclayton @bparees @danmcp I moved the gists to a PR so people can comment and send PR's with changes (and comment on those changes ;-). 

Previous Clayton's questions (copied):

>  Define image repositories as well here

You mean PEP [Image Repository](https://github.com/openshift/openshift-pep/blob/master/openshift-pep-013-openshift-3.md#image-repository) ? Will that repository be defined in JSON by the `imageRegistryURL`? Do we need to have the set of metadata (`metadata: [ {...} ]`) and if so, do you have any examples of the metadata in your mind? (i can think of a 'tag' (?))

> Need to have a discussion about whether build configs and image repositories should be uniquely identified, or a getOrCreate call.

The work flow in my mind is:  The `service` references `image` which is coupled with the `buildConfig/image`. If the Docker image does not exists on minion, then Docker will pull it ('get') from the image repository. If the image is not available in the image repository, it will be built and then pushed into image repository specified in the `buildConfig`.

> Speak with Cesar and Abhisheck but a podTemplate is probably a child of a deployment which is a child of a deployment config, which is a child of a service. We might be able to collapse a deployment config and a service, but not sure

@abhgupta ^^

>  Can we track what env is exported via the links instead?

We can and actually I like it as it is more descriptive (what is exported to where). However, exporting between projects might be tricky, because perhaps you want to restrict what vars will be exported from one project to another.

> Is "from" an id or name? Is there a use case where you need to link across projects? Or not?

The "from" is service id. In new schema I call it "service": "ID". The use-case is right there, you have redis-cluster project, with several services you want to handle independently from your application project. Then you want to link these two together. Alternatively you can have N other projects, you wanna sync via Redis cluster. @bparees ?

mfojtik committed:
Added three example project templates

mfojtik commented:
@smarterclayton @bparees I have question about the [Source Repository](https://github.com/openshift/openshift-pep/blob/master/openshift-pep-013-openshift-3.md#source-code-repository). This will need to be defined for the `buildConfig` (STI configuration) or for the `service` ? From the PEP, it should be part of the `service` configuration, should be optional and should have type "external/internal"?

Looking at this: https://github.com/openshift/openshift-pep/blob/master/openshift-pep-013-openshift-3.md#image-repository

Seems like I need to add this into the `service` description:

``` json
{
    "services": [
        {
            "sourceRepository": {
                "url": "git://github.com/mfojtik/sample-app",
                "type": "external"
            }
        }
    ]
}
```

smarterclayton commented:
Re source: Source repository is not tied to a service.  It's really its own object in a project, linked to a service.  And a build config doesn't necessarily have to depend on a source repository object.  Let's leave this as "undefined" until we can get a better definition.  I don't think we know whether it belongs to service, build, or project yet.

> The work flow in my mind is: The service references image which is coupled with the buildConfig/image. If the Docker image does not exists on minion, then Docker will pull it ('get') from the image repository. If the image is not available in the image repository, it will be built and then pushed into image repository specified in the buildConfig.

Services reference deployment configs, which reference deployments, which reference pod templates, which _can be_ generated from an image.  However, for a template, if you define an image repository and then reference that image repository by name from within a pod template, there's extra work that should happen in the template resolution to replace that reference prior to the actual creation of the service.  Think of the podTemplate as being the thing that is getting replaced with specific variables.

VojtechVitek commented:
Term `service` (as defined in [PEP 13](https://github.com/openshift/openshift-pep/blob/master/openshift-pep-013-openshift-3.md#service)) overlaps with Kubernetes `service` (from [DESIGN #labels](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/DESIGN.md#labels)). There was already some discussion on changing this to `lbservices` or something. So, are we going to use a different name to prevent further confusion?

@smarterclayton

mfojtik commented:
@smarterclayton gotcha, will skip the sourceRepository for now.

Should the `service` have then:

``` json
{
  "services": [
    {
      "id": "service1",
      "name": "service-name",
      "deploymentConfig": { "id": "http://apiserver/deploymentConfigs/123" },
      "env": [ "DB_PASSWORD", "DB_USER" ]
    }
  ]
}
```

And omit the `podTemplate` definition entirely in service?

smarterclayton commented:
> You mean PEP Image Repository ? Will that repository be defined in JSON by the imageRegistryURL? Do we need to have the set of metadata (metadata: [ {...} ]) and if so, do you have any examples of the metadata in your mind? (i can think of a 'tag' (?))

It's unclear, it needs more thought.  For now, assume that a pod template references an image by name - hypothetically if that name mapped to an imageRepository in the template then the name would be replaced by a reference to the latest image from whatever tag was specified.

smarterclayton commented:
> Should I update the JSON to define the relation you described? service -> deploymentConfig -> deployment -> podTemplate.

Yeah, I would prefer it.  Speak with @csrwng about the deploymentConfig object.

smarterclayton commented:
> The "from" is service id. In new schema I call it "service": "ID". The use-case is right there, you have redis-cluster project, with several services you want to handle independently from your application project. 

Can you merge what you've described with what cesar has?

smarterclayton commented:
@VojtechVitek let's continue to call them services for now, the services in kubernetes need to evolve anyway.  The end goal is two different names, we're just not there yet.

mfojtik commented:
> Can you merge what you've described with what cesar has?

Merged, the final form is:

``` json
{
  "serviceLinks": [
    {
      "id": "53c4249f076573c0f4000001",
      "policy": "round-robin",
      "from": [
        {
          "serviceId": "database",
          "export": [
            "DB_NAME",
            "DB_USER",
            "DB_PASSWORD"
          ]
        }
      ],
      "to": [
        {
          "serviceId": "frontend",
          "port": 5432
        }
      ]
    }
 ]
}
```

smarterclayton commented:
Policy is too generic for now - it should be loadBalancerPolicy

smarterclayton commented:
Need to capture mapping of env var names from source to destination somehow.  That might just be part of the service, not the from

bparees commented:
@mfojtik I believe service definitions should still contain a pod definition(template).

mfojtik commented:
@smarterclayton s/policy/loadBalancerPolicy/

smarterclayton commented:
@bparees they do, it should be all the way under deployment config / deployment

bparees commented:
@smarterclayton re: env variable mapping, I believe the idea is that the from/to is the mapping ("map the env variables "from" exposes, into the containers created by the pod corresponding to the service in "to"

csrwng commented:
@mfojtik @smarterclayton - I think we're missing the concept of a template for a pod template :-) 
So the deployment object is a runtime thing which has a limited lifespan and it will have a template for a pod with specific image versions to create. However, that template had to be instantiated from somewhere. We can either store it in the deployment config or the service itself. For now I was thinking it can go in the deployment config

bparees commented:
@smarterclayton @csrwng I was also thinking "deployment" was something that gets generated from a project template, not something the project template defines.

mfojtik commented:
@smarterclayton re: mapping:

I thought we want to handle them in the links https://gist.github.com/mfojtik/6d8b9eb57fb28e3f5ffc#comment-1276093

smarterclayton commented:
@bparees the less generation we do in the template, the better.  Generating a template is a different use case.

bparees commented:
@smarterclayton i'm not sure whether you're agreeing w/ me or disagreeing.  I certainly wasn't talking about generating a project template....

smarterclayton commented:
Regarding deployment - the first time I create a service, will it be required for it to have a deployment.  To create that deployment, someone has to do generation.  Doing the generation when creating the template is a better place than _inside_ the template.  Three phases:

```
define a template (might include generation) -> specialize a template (replace vars and image names) -> create template from specialization
```

W're only solving about specialization here - specialization doesn't need to generate a pod template from an image, so having the final object structure of the template as close as possible to what gets deployed is ideal (less things that have to be done in the template).

csrwng commented:
@bparees - a deployment will be generated from a deployment config

smarterclayton commented:
> re: mapping: I thought we want to handle them in the links

I'm suggesting this:

```
{
  "fromServiceId": "foo",
  "toServiceId": "bar",
  "env": [
    {
      "name": "DB_PASSWORD",
      "toName": "MYSQL_DB_PASS", // optional
  ],
}
```

mfojtik commented:
@csrwng @smarterclayton so project template will be more less like a all-in-one thing that aggregates all components in the 'expanded' way? Eg. you specify the full deploymentConfig under service in project template, then once you submit the template, we go create/link all things together.

bparees commented:
@smarterclayton now you've tied env vars to links, instead of to services, i'm not a fan of that.

bparees commented:
@smarterclayton what would it look like for a project template with 3 interlinked services?

csrwng commented:
@bparees - ultimately you'll get env vars from both -- service definition and links. If you no longer need a link, there is no reason the related env vars should still be around

mfojtik commented:
@bparees `serviceLinks/to` is an array, you can link 3 services together, however the JSON will grow a bit ;-)

smarterclayton commented:
@bparees this is exposure of env vars, not generation.  Links are the conduit through which env var flows.  They are also a mapping.

mfojtik commented:
@smarterclayton https://github.com/mfojtik/oo-api-v3/commit/2e7164c1a1b6cf1c3c95fe4a2a83998829a97cfb#diff-f0be029350fec2851a03e72b6d2b0757R45

csrwng commented:
> @csrwng @smarterclayton so project template will be more less like a all-in-one thing that aggregates all components in the 'expanded' way? Eg. you specify the full deploymentConfig under service in project template, then once you submit the template, we go create/link all things togeth

Yes, my thinking is that templates should not depend on any resources that may/may not exist.

bparees commented:
@mfojtik what if I want to like A->B and B->C and C->A?

mfojtik commented:
@bparees:

``` json
{
  "serviceLinks": [
    {
      "id": "53c4249f076573c0f4000001",
      "loadBalancerPolicy": "round-robin",
      "from": [
        {
          "serviceId": "A",
        }
      ],
      "to": [
        {
          "serviceId": "B",
        }
      ]
    }.
    {
      "id": "53c4249f076573c0f4000002",
      "loadBalancerPolicy": "round-robin",
      "from": [
        {
          "serviceId": "B",
        }
      ],
      "to": [
        {
          "serviceId": "C",
        }
      ]
    },
    {
      "id": "53c4249f076573c0f4000003",
      "loadBalancerPolicy": "round-robin",
      "from": [
        {
          "serviceId": "C",
        }
      ],
      "to": [
        {
          "serviceId": "A",
        }
      ]
    }
 ]
}
```

bparees commented:
@smarterclayton yeah, so i have to list the env vars twice (once in the service that defines them, and once in the link that transports them).  We were just going to (for now) skip defining which get transported, and just transport all of them when a link exists.

smarterclayton commented:
@bparees maybe make that explicit (export all)

smarterclayton commented:
@mfojtik instead of link/from/env, do link/env.

mfojtik commented:
@smarterclayton updated: https://github.com/mfojtik/oo-api-v3/commit/ce890fa07c709a52cdce8e19feebdcb316f49b36#diff-f0be029350fec2851a03e72b6d2b0757R40

mfojtik commented on examples/template_example_1.json:
```diff
@@ -0,0 +1,119 @@
+        },
+        {
+          "name": "DB_PASSWORD",
+          "to_name": "MYSQL_ADMIN_PASSWORD"
```
ntm, typo

mfojtik commented:
@bparees @smarterclayton maybe export all by default (if not listed what to export).

mfojtik commented:
@bparees , @smarterclayton , @csrwng 

**Summary of changes:**
- Added `imageRepository` definition (_id_, _url_, _name_)
- Added references from `imageRepository` to `buildConfig` and `podTemplate`
- Added image `tag` to `podTemplate` image definition
- Added `description` to project and to services
- Updated `serviceLinks` `from` and `to` to be a hash instead of array
- The `port` in serviceLinks renamed to `ports` and converted to an array (exporting multiple ports)
- Wrapped `podTemplate` in `deploymentConfig/deployment`
- Copied `replicas`, `replicaSelector` and `replicaPolicy` to example2 (redis-slave)

mfojtik commented:
@csrwng ^^

csrwng commented:
@mfojtik @VojtechVitek we need a flag to specify whether the deployment is kicked off automatically on image repository updates.

smarterclayton commented:
Re: export all if none specified - i don't think the only purpose of a link is to expose environment variables in the long term.  So we need to distinguish between none, all, and some.

smarterclayton commented:
re image updates - each source image repository is a trigger for a deployment - so the flag is on a set of image repositories.

smarterclayton commented:
re: image references - it's probably better to do string substitution in a template for the first deployment which is "use the image from this repository with this tag as the value for podTemplate.image"

mfojtik commented:
@smarterclayton so if 'all' is the default, then none will be `env: []`

mfojtik commented:
@smarterclayton image references: JSON updated

mfojtik commented:
@smarterclayton @csrwng @bparees can we do some final review for the JSON files and then merge them? Would be easier then do propose changes/updates than updating this PR ;-)

smarterclayton commented on examples/template_example_1.json:
```diff
@@ -0,0 +1,132 @@
+  "kind": "Project",
+  "buildConfig": [
+    {
+      "sourceUri": "https://raw.githubusercontent.com/mfojtik/phpapp/master/Dockerfile",
```
build configs need a name

smarterclayton commented on examples/template_example_1.json:
```diff
@@ -0,0 +1,132 @@
+    {
+      "imageRepositoryId": "53c4249f076573c0f4000046",
+      "sourceUri": "https://raw.githubusercontent.com/docker-library/postgres/docker/9.2/Dockerfile",
+      "type": "docker"
```
I'm uncertain about type yet, but it's fine for now.  Have andy and them signed off yet?

mfojtik replied:
the 'docker' type is what they have in their PR, i'm not sure about the other types, will sync with them

smarterclayton commented on examples/template_example_1.json:
```diff
@@ -0,0 +1,132 @@
+    }
+  ],
+  "description": "Example PHP application with PostgreSQL database",
+  "envTemplate": [
```
I'm wondering whether we should be doing this on ENV, or just on free parameters that can be used anywhere.  For instance, what if need to parameterize something that's not an env var?

mfojtik replied:
just to be clear, you mean use `envTemplate` items as a template variables, eg. use `$DB_PASSWORD` anywhere in the template JSON and it will be substituted by the generated value?

i will rename the `envTemplate` then, to something like `variables`.

smarterclayton replied:
Yeah - have variables be broader.  I'd recommend checking out the HOT specification (used by heat in openstack) for terminology we can align with.  We are not trying to replicate the complexity of HOT, just be compatible in terminology

smarterclayton commented on examples/template_example_1.json:
```diff
@@ -0,0 +1,132 @@
+      "type": "docker"
+    },
+    {
+      "imageRepositoryId": "53c4249f076573c0f4000046",
```
imageRepositories are referenced by "name" as their ID for now - so would be "servername/username/reponame" - use `postgres` here for now.

smarterclayton commented on examples/template_example_1.json:
```diff
@@ -0,0 +1,132 @@
+  "name": "my-awesome-php-app",
+  "serviceLinks": [
+    {
+      "id": "53c4249f076573c0f4000001",
```
Drop IDs on links here.

bparees replied:
@smarterclayton @mfojtik per the discussion this afternoon, it sounds like links are being dropped entirely.

@mfojtik I still think you should merge what you've got and we can evolve, but the new idea as I understand it is to simply define all env variables at the project level, and then reference them from anywhere you want them to be available (ie both a ruby container and a postgres container would have a reference to $DB_PASSWORD, and the project itself would define DB_PASSWORD=generate()  (syntax for example purposes only))

network linking between containers (and injecting endpoint information for remote containers would be handled by defining what k8s calls a service (what i'd rather call an endpoint or endpointcluster) and anything that wants to use it (eg a ruby deployment wanting to use postgres) would include a reference to the service/endpoint/endpointcluster definition.

@smarterclayton that about right?

mfojtik replied:
@bparees 'service' in k8s right now is not isolated, iow. you define a service and then it is available to _all_ pods (not just to 'project' pods/services...).

I will understand dropping the links entirely, if the k8s service/endpoint definition include target/destination openshift service.

mfojtik replied:
@bparees re env vars: I think we are already there, `envTemplate` define vars at project level and then we reference in podTemplate  or links, etc.. @smarterclayton suggested to put them into container definition where they belong, my example was that if you put them into podTemplate, they will be shared by **all** containers within the Pod.

bparees replied:
@mfojtik the ability to put them in a pod rather than having to put them in each container individually sounds like a convenience but not a requirement for functionality.  You could argue for a similar ability to define the variable at the project level and have them put into all containers of all pods in the project, right?  

I guess I don't have a strong opinion either way, but I have a slight preference for the simplicity of just saying "if you want the variable available to the container, put it in the container definition" for now.

smarterclayton commented on examples/template_example_1.json:
```diff
@@ -0,0 +1,132 @@
+        "deployment": {
+          "podTemplate": {
+            "env": [
+              "DB_PASSWORD",
```
This isn't how env is set in real pod templates.  It needs to match 1-1.  If that means you need to define some sort of replacement policy, so be it.

smarterclayton replied:
Also, pods don't have env, containers have env.

mfojtik replied:
Idea was that if you define env vars on podTemplate level, then they are shared by all containers inside the pod.

smarterclayton commented on examples/template_example_1.json:
```diff
@@ -0,0 +1,132 @@
+            "containers": [
+              {
+                "hooks": {
+                  "prestart": {
```
this is really more of the deployment job description.  speak with cesar and get an example of how he's actually defining the type of deployment process to be run.

mfojtik replied:
@csrwng ^^

smarterclayton commented on examples/template_example_1.json:
```diff
@@ -0,0 +1,132 @@
+  "envTemplate": [
+    {
+      "generate": "[a-zA-Z0-9]{8}",
+      "id": "DB_PASSWORD",
```
Use "name" instead of "id"

mfojtik commented:
@smarterclayton I made all changes you proposed, I renamed `envTemplate` to `parameters` to be aligned with Heat HOT, then items defined inside parameters could be used for replacements in `env` (moved from services to `containers` (btw. we need to update PEP for this)).

mfojtik commented:
@smarterclayton btw. I have two questions:
- Do you have an example how you can define ENV vars in kubernetes Pod description? I used key=>value as I was not able to find any example so far.
- The `labels` vs. name. Do you think we still do need `name` when we have `labels : { "name": "foo"}` ?

VojtechVitek commented:
> The labels vs. name. Do you think we still do need name when we have labels : { "name": "foo"} ?
> @mfojtik @smarterclayton If we use labels, should those be tied with services, pods or both?

mfojtik commented:
@VojtechVitek my question was more about if the 'name' and 'labels>name' are orthogonal or they are basically duplicate information.

smarterclayton commented:
Labels are a user construct - users should own their definition completely - and so a label for name is not appropriate

smarterclayton commented:
@mfojtik re: env vars they're defined in containers.  for this work the only place where they need to be are in the container in the template.

VojtechVitek commented:
E-R diagram v2: https://docs.google.com/a/redhat.com/document/d/1aTrsQnu0IsBNGVz5GoUmVqdD0zQHclYU5rcCiU1VLeg

mfojtik commented on examples/template_example_1.json:
```diff
@@ -0,0 +1,143 @@
+      "description": "PostgreSQL username"
+    },
+    {
+      "type": "string",
```
the type could be 'uuid', 'password', etc... see the information sharing docs.

smarterclayton replied:
let's spec type as only "string" for now.

smarterclayton replied:
also, in the examples, order them name, description, type, generate

smarterclayton commented on examples/template_example_1.json:
```diff
@@ -0,0 +1,143 @@
+      "url": "registry.hub.docker.com/postgres"
+    }
+  ],
+  "kind": "Project",
```
Nuke this - I don't think we want it in the template for now

smarterclayton commented on examples/template_example_1.json:
```diff
@@ -0,0 +1,143 @@
+    }
+  ],
+  "kind": "Project",
+  "name": "my-awesome-php-app",
```
move name, id, and description to the top so it's easy to read

smarterclayton commented on examples/template_example_1.json:
```diff
@@ -0,0 +1,143 @@
+  ],
+  "serviceLinks": [
+    {
+      "loadBalancerPolicy": "round-robin",
```
let's just drop this from the template for now

smarterclayton commented on examples/template_example_1.json:
```diff
@@ -0,0 +1,143 @@
+      "id": "database",
+      "name": "database",
+      "description": "Standalone PostgreSQL 9.2 database service",
+      "type": "stateful",
```
nuke this

smarterclayton commented on examples/template_example_1.json:
```diff
@@ -0,0 +1,143 @@
+  ],
+  "services": [
+    {
+      "id": "database",
```
I'd remove ID from this, and define links as working on names for now.  The reason being in the template that IDs will be late generated (potentially on the server) and it's not clear to me how they'll work once we have updates (we need to update by something that is != a generated ID).

mfojtik commented:
@smarterclayton fixed
